marshmallow_dataclass/generic_resolver.py

```diff
@@ -0,0 +1,193 @@
+import copy
+import dataclasses
+import inspect
+import sys
+from typing import (
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Tuple,
+    TypeVar,
+)
+
+import typing_extensions
+import typing_inspect
+
+if sys.version_info >= (3, 9):
+    from typing import Annotated
+else:
+    from typing_extensions import Annotated
+
+_U = TypeVar("_U")
+
+
+class UnboundTypeVarError(TypeError):
+    """TypeVar instance can not be resolved to a type spec.
+
+    This exception is raised when an unbound TypeVar is encountered.
+
+    """
+
+
+class InvalidStateError(Exception):
+    """Raised when an operation is performed on a future that is not
+    allowed in the current state.
+    """
+
+
+class _Future(Generic[_U]):
+    """The _Future class allows deferred access to a result that is not
+    yet available.
+    """
+
+    _done: bool
+    _result: _U
+
+    def __init__(self) -> None:
+        self._done = False
+
+    def done(self) -> bool:
+        """Return ``True`` if the value is available"""
+        return self._done
+
+    def result(self) -> _U:
+        """Return the deferred value.
+
+        Raises ``InvalidStateError`` if the value has not been set.
+        """
+        if self.done():
+            return self._result
+        raise InvalidStateError("result has not been set")
+
+    def set_result(self, result: _U) -> None:
+        if self.done():
+            raise InvalidStateError("result has already been set")
+        self._result = result
+        self._done = True
+
+
+def is_generic_alias(clazz: type) -> bool:
+    """
+    Check if given class is a generic alias of a class is
+    defined as `class A(Generic[T])`, this method will return true if `A[int]` is passed
+    """
+    is_generic = is_generic_type(clazz)
+    type_arguments = typing_extensions.get_args(clazz)
```

Comment:
Can we just use `typing.get_args` here?

Then, just one-line this, with short-circuit:

```py
return is_generic_type(clazz) and len(get_args(clazz)) > 0
```

Reply:
In general, can we use `typing.get_args` in place of `typing_extensions.get_args` (and `typing.get_origin` in place of typing_extensions.get_origin`) everywhere?

Both `get_origin` and `get_args` are in `typing` as of python 3.8.

Reply:
They are both in 3.8, but when I use the `typing` functions the tests break. The implementation appears incomplete.
From 3.9 onwards the `typing` functions work fine.

Reply:
Now I'm curious as to what the difference is.

What happens if one just accesses `.__origin__` and `.__args__` directly?  ([ref](https://docs.python.org/3/library/stdtypes.html#genericalias.__origin__))

Reply:
I attributed the failed test to the wrong root cause.

Py 3.8 doesn't support Annotated, so we import it from `typing_extension` Therefore we need to use `get_origin` and `get_args` from typing_extension as well.

Reply:
We could do this:
```python
if sys.version_info >= (3, 9):
    from typing import Annotated, get_args, get_origin
else:
    from typing_extensions import Annotated, get_args, get_origin
```

And just use the function without module in the code.

Reply:
> Py 3.8 doesn't support Annotated, so we import it from `typing_extension` Therefore we need to use `get_origin` and `get_args` from typing_extension as well.

So it only matters when dealing with `Annotated`s?

Reply:
Correct.